### PR TITLE
Refs #6763 Feature/xtypes dds

### PR DESCRIPTION
**Pull request by LuisGP**
- [x] Migrate method signatures to use `xtypes::DynamicData` instead of `soss::Message`.
- [x] Implement conversion methods between `xtypes::DynamicData` and `eprosima::fastrtps::types::DynamicData`.
- [x] Uncomment and fix implementation of SOSS-DDS methods.
- [x] Easy FastRPTS installation and install documentation
- [ ] https://github.com/osrf/soss/pull/20 merged

### CODE REVIEW

File: dds-test/src/integration/roundtrip.cpp [hunk 1]
```diff
@@ -149,11 +148,16 @@ TEST_CASE("Transmit to and receive from dds", "[dds]")
...
+            msg_to_sent["data"].value(message_data);
+
             // Road: [mock -> dds -> dds -> mock]
-            soss::Message msg_to_recv = roundtrip(topic_sent, topic_recv, msg_to_sent);
+            xtypes::DynamicData msg_to_recv = roundtrip(topic_sent, topic_recv, msg_to_sent);
 
-            REQUIRE(*msg_to_sent.data.at("data").cast<std::string>() == *msg_to_recv.data.at("data").cast<std::string>());
-            REQUIRE(msg_to_sent.type == msg_to_recv.type);
+            REQUIRE(msg_to_sent == msg_to_recv);
+            // TODO
```
Comment (lemunozm) [inline dds-test/src/integration/roundtrip.cpp]: It is still a TODO? (see similar occurrences)

File: dds/colcon.pkg [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (lemunozm) [inline dds/colcon.pkg]: Is this file really necessary?

File: dds/doc/diagram-design/class.puml [hunk 3]
```diff
@@ -31,7 +31,7 @@ class Publisher {
     - dds::dynamic_data
     - topic_name
     - message_type
-    + publish(soss_message)
+    + publish(xtypes)
```
Comment (lemunozm) [inline dds/doc/diagram-design/class.puml]: I would modify xtypes by DynamicData in this context (see similar occurrences)

File: dds/src/Conversion.hpp [hunk 4]
```diff
@@ -21,22 +21,85 @@
...
+            DynamicPubSubType* type)
+    {
+        registered_types_.emplace(type_name, type);
+    }
+
+    static DynamicTypeBuilder* create_builder(
+            const xtypes::DynamicType& type);
 private:
     ~Conversion() = default;
+    static std::map<std::string, ::xtypes::DynamicType::Ptr> types_;
```
Comment (lemunozm) [inline dds/src/Conversion.hpp]: What happen if there is two soss-dds? could this map be shared between them and cause problems?

File: dds/src/Subscriber.cpp [hunk 5]
```diff
@@ -36,20 +36,26 @@ namespace dds {
...
         TopicSubscriberSystem::SubscriptionCallback soss_callback)
 
     : topic_name_{topic_name}
     , message_type_{message_type}
     , soss_callback_{soss_callback}
     , reception_threads_{}
 {
-    dynamic_data_ = participant->create_dynamic_data(message_type);
+    DynamicTypeBuilder* builder = Conversion::create_builder(message_type);
+    if (builder != nullptr)
```
Comment (lemunozm) [inline dds/src/Subscriber.cpp]: print an error if not? in which cases could the builder is null?

File: dds/src/SystemHandle.cpp [hunk 6]
```diff
@@ -34,6 +33,7 @@ namespace dds{
 namespace {
 
 // This function patches the problem of dynamic types, which do not admit '/' in their type name.
+/*
```
Comment (lemunozm) [inline dds/src/SystemHandle.cpp]: This saved us about the type name from ros2 with `/`. How do you solve this?

**Review by lemunozm — CHANGES_REQUESTED** [review]

File: dds-test/src/integration/roundtrip.cpp (hunk 1 above)
Comment (LuisGP) [reply dds-test/src/integration/roundtrip.cpp]: Yes, it depends on the `operator ==` of xtypes.

File: dds/colcon.pkg (hunk 2 above)
Comment (LuisGP) [reply dds/colcon.pkg]: Probably not, but it is recommended when using colcon.

File: dds/src/Conversion.hpp (hunk 4 above)
Comment (LuisGP) [reply dds/src/Conversion.hpp]: As far as the user keeps the mutual exclusion and doesn't reuse type names, it shouldn0t be a problem. Do you think we should protect ourselves against evil users?

File: dds/src/Subscriber.cpp (hunk 5 above)
Comment (LuisGP) [reply dds/src/Subscriber.cpp]: It can be null if the type contains unsupported types, which currently are none (in the other way, from dss to soss, it is possible).
Added a throw, just in case.

File: dds/src/SystemHandle.cpp (hunk 6 above)
Comment (LuisGP) [reply dds/src/SystemHandle.cpp]: I didn't care about it, but I'm sure it is necessary. That's because I leave it commented.

File: dds/colcon.pkg (hunk 2 above)
Comment (lemunozm) [reply dds/colcon.pkg]: I ask because the rest of the packages do not have this file. But it is ok for me to let it alive

File: dds/src/Subscriber.cpp (hunk 5 above)
Comment (lemunozm) [reply dds/src/Subscriber.cpp]: ok, :+1:

File: dds-test/src/integration/roundtrip.cpp (hunk 1 above)
Comment (lemunozm) [reply dds-test/src/integration/roundtrip.cpp]: Ok, we remove the comment and in the future, maybe we can create this operator.

File: dds/src/SystemHandle.cpp (hunk 6 above)
Comment (lemunozm) [reply dds/src/SystemHandle.cpp]: ok

File: dds/src/Conversion.hpp (hunk 4 above)
Comment (lemunozm) [reply dds/src/Conversion.hpp]: spoken in person

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/19/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/26/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/27/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/29/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/32/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/35/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/39/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/50/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/52/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/54/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/56/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss-dds-linux/58/)

**Comment by jamoralp** [comment]
Closing this, since the reworking from SOSS to Integration Service is now near to be completed. `feature/xtypes-dds` branch will be transferred to `main`, and then the code migration from `soss` to `eprosima::is` in the namespaces, as well as using Fast-DDS 2.X.X new API will be done in a separate pull request.
